### PR TITLE
Added support for evolutionary models

### PR DESCRIPTION
**Pull request by bennati**
As discussed in Issue #359 , I implemented some new classes that allow to simplify the creation of evolutionary models, i.e. where agents accumulate energy, reproduce, mutate and die.

In order to show how to use the model, I refactored the wolf-sheep example to use the new classes.
The example does not use all features of the new classes, so if you have another evolutionary model in mind I could implement it as an example.

This code is to be considered preliminary, I'm open to discussion about how to improve its design to make it more general and about implementation of new (selection) algorithms.

Cheers,
Stefano

### CODE REVIEW

**Comment by dmasad** [comment]
Hi Stefano,

This looks great! I think you've built a really cool set of components, and I want to get a chance to review further and suggest tweaks or improvements.

My main concern is that I'm not sure whether a specific evolutionary model is a good fit for adding to the Mesa core, since it does currently have a lot of hard-coded assumptions and a limited number of example cases. 

One thing we've talked about doing before is Mesa kits -- extensions to the core framework which add specific features like geography, or -- in this case -- evolution. What are your thoughts on using this as a first example of that, and spinning this off into a `mesa.evolution` or something like that sub-module?

**Comment by bennati** [comment]
Sure, it sounds like a good idea! Will you take care of the changes?
Please discard my last commit as that is a change in the core module, I'll submit a separate pull request for that later

File: mesa/EvoAgent.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+    def step(self):
+        """
+        Performs one step of simulation
+
+        Side effect: age is increased by 1
+        """
+        self.age+=1
+
+    def duplicate(self):
```
Comment (rht) [inline mesa/EvoAgent.py]: I couldn't tell why you chose not to use "clone", given the domain area of the model ;)

Comment (rht) [inline mesa/EvoAgent.py]: Or rather, clone in the strictest sense is DNA->DNA copy, but duplicate is for molecule->molecule copy. The former is just the recipe, which doesn't even include the epigenome, the latter's definition includes all that is necessary.

File: examples/wolf_sheep/wolf_sheep/agents.py [hunk 2]
```diff
@@ -1,6 +1,5 @@
 import random
-
-from mesa import Agent
+from mesa.EvoAgent import *
```
Comment (rht) [inline examples/wolf_sheep/wolf_sheep/agents.py]: `import *` makes it harder for `pylint`/`flake8` to statically analyze the code.

**Comment by rht** [comment]
This fixes #359 if nailed (to keep track of the things...).

**Comment by bennati** [comment]
Hi guys, sorry about the long absence.
Could you please explain what I have to do in order to make this a sub-module? Is there any example/guide?
Thank you!

**Comment by jackiekazil** [comment]
@bennati -- you bring up a great point. It is not clear how this might move forward design wise. Since everyone doesn't follow every ticket, I have posted to the dev email list. Hopefully, we can get some folks to chime in there. See: https://groups.google.com/forum/#!topic/projectmesa-dev/4HThxwC5sGQ 

For now, I am going to consider this 'blocked'. Hopefully we can get you unblocked soon!

**Comment by jackiekazil** [comment]
@bennati ... looks like we are narrowing in on a model for this. I will update when have something solidified. -- https://groups.google.com/forum/#!topic/projectmesa-dev/4HThxwC5sGQ

**Comment by jackiekazil** [comment]
@bennati after meeting about this, we think the best way to approach this is to create a separate library that works with Mesa. For example - evolutionary-mesa or evol-mesa or mesa-evol or something like that. That library would have mesa w/ a specific version number as a requirement. This would be hosted under your account or another organization. We would point people to it as a resource. Over time, if it became a key component with wide adoption, we might discuss hosting it under the Mesa org, but still keeping it separate from core mesa. We will also be here for you a resource to help you structure the library correctly. When you start a repo, please post it here or in a ticket, so that we can poke around and help you get up and running.

**Pull request closed without merging**